pyanaconda/dbus/errors.py

```diff
@@ -0,0 +1,64 @@
+#
+# Known DBus errors.
+#
+# Copyright (C) 2018  Red Hat, Inc.  All rights reserved.
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+from pydbus.error import map_error, map_by_default
+
+from pyanaconda.dbus.constants import DBUS_BOSS_ANACONDA_NAME, DBUS_BOSS_INSTALLATION_NAME, \
+    DBUS_TASK_NAME, ANACONDA_DBUS_NAMESPACE
+
+
+@map_by_default
+class DBusError(Exception):
+    """A default DBus error."""
+    pass
+
+
+@map_error("{}.Error".format(ANACONDA_DBUS_NAMESPACE))
+class AnacondaError(Exception):
+    """A default DBus error."""
+    pass
+
+
+@map_error("{}.SplitKickstartError".format(DBUS_BOSS_ANACONDA_NAME))
+class SplitKickstartError(AnacondaError):
+    """Error while parsing kickstart for splitting."""
+    pass
+
+
+@map_error("{}.SplitKickstartSectionParsingError".format(DBUS_BOSS_ANACONDA_NAME))
+class SplitKickstartSectionParsingError(SplitKickstartError):
+    """Error while parsing a section in kickstart."""
+    pass
+
+
+@map_error("{}.SplitKickstartMissingIncludeError".format(DBUS_BOSS_ANACONDA_NAME))
+class SplitKickstartMissingIncludeError(SplitKickstartError):
+    """File included in kickstart was not found."""
+    pass
+
+
+@map_error("{}.InstallationNotRunning".format(DBUS_BOSS_INSTALLATION_NAME))
+class InstallationNotRunning(AnacondaError):
+    """Exception will be raised when action requires running installation."""
+    pass
+
+
+@map_error("{}.AlreadyRunning".format(DBUS_TASK_NAME))
+class TaskAlreadyRunningException(AnacondaError):
+    """Exception will be raised when starting task which is already running."""
+    pass
```

Comment:
I really don't think this is the correct solution. I agree that we need to have exceptions on place, everyone would know where to look than. On the other hand you don't want to have here exceptions specific for modules (e.g. StorageDeviceFormatingError) it would be problem even because of naming.

I would propose that these common and Boss exceptions should be placed in `pyanaconda/dbus/errors/__init__.py` and every other module will place their exceptions in special file.
So there will be:
`pyanaconda/dbus/errrors/__init__.py`
`pyanaconda/dbus/errors/timezone.py`
`pyanaconda/dbus/errors/payload.py`
`...`

After this you will have all the exceptions on one place but they could be extended by modules and won't be present when this module won't be available in the system.

Reply:
I like your solution, thanks!

Reply:
One note about this. The mapping happens during the import, so if all errors are defined in one file it is enough to import one file to be able to recognize all DBus errors. If we split the errors to multiple files, the client will not able to recognize errors from different modules unless it imports all files.

Reply:
That is true but on the other note we won't be able to easily add new errors to the project. Also how the addons will their errors?
However, do we really need it? When a module wants to communicate with other module then it has that module as dependency and know where to look for their exceptions. I don't see a point in having all these errors imported even errors which can't be raised at all because that error is used by the module which I'm not touching at all.

Reply:
This indeed seems like a tricky issue - suppose I want to handle an exception raised by another module.

If the exceptions are all in one file:
+ easy import
+ not a problem if the module itself is not present
- lot of possibly unrelated stuff in the file
- addons can't put their specific errors there, but that should not be an issue (Boss & Anaconda modules should not care about those)

If we have per module files:
+ less clutter due to separate files
+ addons can drop such a file as well, but that's likely useless (Boss and Anaconda modules should ever care only about generic errors raised by an addons, not any custom ones)
- modules need to explicitly care about other modules, introducing tighter coupling
- if we package modules separately and include this file, there could be crashes at runtime if some modules are not present (eq. no storage module in Initial Setup run)

So I'm not really sure what's the best option - maybe some string matching for custom error instead or something dynamic importless like that ? Also what about hypothetical modules/addons that are not Python based ?